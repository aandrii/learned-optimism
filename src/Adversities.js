import React from 'react';
import ReactFireMixin from 'reactfire';
import { Button, FormControl, Form, FormGroup, InputGroup } from 'react-bootstrap';
import { Link, Redirect } from 'react-router-dom';
import Spinner from 'react-spinner';
import firebase from 'firebase';

import userRefFor from './userRef'

export default React.createClass({
  mixins: [ReactFireMixin],
  propTypes: {
    user: React.PropTypes.instanceOf(firebase.User).isRequired
  },
  getInitialState() {
    return {
      description: ''
    };
  },
  componentWillMount() {
    this.bindAsArray(userRefFor(this.props.user).child('adversities'), 'adversities');    
    // Once the data has loaded for the first time, stop displaying the spinner
    this.firebaseRefs.adversities.once('value').then(() => this.setState({loaded: true}));
  },
  render() {
    const {adversities, newAdversityId} = this.state;

    return (
      this.state.loaded ? 
        // If we've just created a new adversity
        newAdversityId ?
          // Redirect to it
          <Redirect to={`/adversities/${newAdversityId}`}/>
          :
          // Otherwise just display the regular form
          <div>
            <Form onSubmit={this.handleSubmit}>
              <FormGroup>
                <InputGroup>
                  <FormControl type='text' 
                         placeholder='Enter todo here' 
                         value={this.state.description}
                         onChange={this.handleChange}/>
                  <InputGroup.Button>
                    <Button type="submit" disabled={this.state.isSaving}>Add</Button>
                  </InputGroup.Button>
                </InputGroup>
              </FormGroup>
            </Form>
            <div className="list-group">
              {adversities.map(adversity => {
                const id = adversity['.key']; 

                return (
                  <Link key={id} className="adversity list-group-item" to={`/todos/${id}`}>
                    {adversity.description}
                    <span onClick={e => {
                            e.preventDefault();
                            this.remove(id);
                          }}
                          className="remove glyphicon glyphicon-remove"/>
                  </Link>
                );
              })}
            </div>
          </div>
        : 
        <Spinner/>
    );
  },
  handleChange(e) {
    e.preventDefault();
    this.setState({description: e.target.value});
  },
  handleSubmit(e) {
    e.preventDefault();
    this.setState({isSaving: true});
    this.firebaseRefs.adversities.push({
      description: this.state.description
    }).then(adversity => {
      this.setState({newAdversityId: adversity.key});
    });
  },
  /**
   * Remove an adversity and all of its associated beliefs
   */
  remove(adversityId) {
    // TODO See if there's a way to do this in a transaction
    this.firebaseRefs.adversities.child(adversityId).remove().then(() => {
      // Delete the beliefs for the adversity
      // TODO See if there's a way to do this in one hit rather than iteratively
      return userRefFor(this.props.user).child('beliefs').orderByChild('adversityId').equalTo(adversityId)
              .once("value").then(snapshot => {
        let promises = [];

        snapshot.forEach(childSnapshot => {
          promises = promises.concat(childSnapshot.ref.remove());
        });

        return Promise.all(promises);
      });
    }).catch(error => console.error(error));
  }
});
