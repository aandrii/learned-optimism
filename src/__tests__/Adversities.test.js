import React from "react";
import { Presentation } from "../Adversities";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const tree = renderer.create(<Presentation />).toJSON();
  expect(tree).toMatchSnapshot();
});
