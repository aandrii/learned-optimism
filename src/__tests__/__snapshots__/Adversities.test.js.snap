exports[`test renders correctly 1`] = `
<div
  className=" react-spinner">
  <div
    className="react-spinner_bar"
    style={
      Object {
        "WebkitAnimationDelay": "-1.2s",
        "WebkitTransform": "rotate(0deg) translate(146%)",
        "animationDelay": "-1.2s",
        "transform": "rotate(0deg) translate(146%)",
      }
    } />
  <div
    className="react-spinner_bar"
    style={
      Object {
        "WebkitAnimationDelay": "-1.1s",
        "WebkitTransform": "rotate(30deg) translate(146%)",
        "animationDelay": "-1.1s",
        "transform": "rotate(30deg) translate(146%)",
      }
    } />
  <div
    className="react-spinner_bar"
    style={
      Object {
        "WebkitAnimationDelay": "-1s",
        "WebkitTransform": "rotate(60deg) translate(146%)",
        "animationDelay": "-1s",
        "transform": "rotate(60deg) translate(146%)",
      }
    } />
  <div
    className="react-spinner_bar"
    style={
      Object {
        "WebkitAnimationDelay": "-0.9s",
        "WebkitTransform": "rotate(90deg) translate(146%)",
        "animationDelay": "-0.9s",
        "transform": "rotate(90deg) translate(146%)",
      }
    } />
  <div
    className="react-spinner_bar"
    style={
      Object {
        "WebkitAnimationDelay": "-0.8s",
        "WebkitTransform": "rotate(120deg) translate(146%)",
        "animationDelay": "-0.8s",
        "transform": "rotate(120deg) translate(146%)",
      }
    } />
  <div
    className="react-spinner_bar"
    style={
      Object {
        "WebkitAnimationDelay": "-0.7s",
        "WebkitTransform": "rotate(150deg) translate(146%)",
        "animationDelay": "-0.7s",
        "transform": "rotate(150deg) translate(146%)",
      }
    } />
  <div
    className="react-spinner_bar"
    style={
      Object {
        "WebkitAnimationDelay": "-0.6s",
        "WebkitTransform": "rotate(180deg) translate(146%)",
        "animationDelay": "-0.6s",
        "transform": "rotate(180deg) translate(146%)",
      }
    } />
  <div
    className="react-spinner_bar"
    style={
      Object {
        "WebkitAnimationDelay": "-0.5s",
        "WebkitTransform": "rotate(210deg) translate(146%)",
        "animationDelay": "-0.5s",
        "transform": "rotate(210deg) translate(146%)",
      }
    } />
  <div
    className="react-spinner_bar"
    style={
      Object {
        "WebkitAnimationDelay": "-0.4s",
        "WebkitTransform": "rotate(240deg) translate(146%)",
        "animationDelay": "-0.4s",
        "transform": "rotate(240deg) translate(146%)",
      }
    } />
  <div
    className="react-spinner_bar"
    style={
      Object {
        "WebkitAnimationDelay": "-0.3s",
        "WebkitTransform": "rotate(270deg) translate(146%)",
        "animationDelay": "-0.3s",
        "transform": "rotate(270deg) translate(146%)",
      }
    } />
  <div
    className="react-spinner_bar"
    style={
      Object {
        "WebkitAnimationDelay": "-0.2s",
        "WebkitTransform": "rotate(300deg) translate(146%)",
        "animationDelay": "-0.2s",
        "transform": "rotate(300deg) translate(146%)",
      }
    } />
  <div
    className="react-spinner_bar"
    style={
      Object {
        "WebkitAnimationDelay": "-0.1s",
        "WebkitTransform": "rotate(330deg) translate(146%)",
        "animationDelay": "-0.1s",
        "transform": "rotate(330deg) translate(146%)",
      }
    } />
</div>
`;
